src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,676 @@
+ *}
+ * <br>
+ * <h2><a id="ConcurrentAccess">Concurrent Access</a></h2>
+ * Unless otherwise documented by an implementation, the methods defined in
+ * this class are not thread-safe. Multiple threads that need to access a
+ * single object concurrently should synchronize amongst themselves and
+ * provide the necessary locking. Multiple threads each manipulating separate
+ * objects need not synchronize.
+ * <br>
+ * <h2><a id="DelayedProviderSelection">Delayed Provider Selection</a></h2>
+ * If a provider is not specified when calling one of the {@code getInstance}
+ * methods, the implementation delays the selection of the provider until the
+ * {@code deriveKey} or {@code deriveData} method is called. This is called
+ * <i>delayed provider selection</i>. The primary reason this is done is to
+ * ensure that the selected provider can handle the key material that is passed
+ * to those methods - for example, the key material may reside on a
+ * hardware device that only a specific {@code KDF} provider can utilize.
+ * <p>
+ * If a provider is not specified in the {@code getInstance} method when
+ * instantiating a {@code KDF} object, the provider is selected the first time
+ * the {@code deriveKey} or {@code deriveData} method is called, and a provider
+ * is chosen that supports the parameters passed to the {@code deriveKey} or
+ * {@code deriveData} method. If the {@code getProviderName} or {@code
+ * getParameters} method is called before the {@code deriveKey} or {@code
+ * deriveData} methods, the first provider supporting the KDF algorithm and
+ * optional {@code KDFParameters} is chosen. This provider may not support
+ * the key material that is subsequently passed to the deriveKey or
+ * deriveData methods. Therefore, it is recommended not to call the {@code
+ * getProviderName} or {@code getParameters} methods until after a key
+ * derivation operation. Once a provider is selected, it cannot be changed.
+ *
+ * @see KDFParameters
+ * @see SecretKey
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public final class KDF {
+    private static final Debug debug = Debug.getInstance("jca", "KDF");
+
+    private static final Debug pdebug = Debug.getInstance("provider",
+                                                          "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=")
+                                             && !Debug.isOn("kdf");
+
+    private record Delegate(KDFSpi spi, Provider provider) {}
+
+    private Delegate pairOfSpiAndProv;
+    private Delegate firstPairOfSpiAndProv;
+
+    // The name of the KDF algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final KDFParameters kdfParameters;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
```

Comment:
It might be better to do the following, as it should be the same thing no matter which constructor is used.

```suggestion
    private final Object lock = new Object();
```